File: game/Rodada.js
// Rodada.js
import { Baralho } from './Baralho.js';
import { Mesa } from './Mesa.js';

export class Rodada {
    constructor(gameSettings) {
        // Puxamos as configurações diretamente da instância do Game que foi passada
        this.gameOrder = gameSettings.gameOrder;
        this.round = gameSettings.round;
        this.randomShuffle = gameSettings.randomShuffle;
        // Mesmo rng do Game (ver Game.rng / rng.js): sem seed é Math.random;
        // com seed é o PRNG determinístico. undefined em quem constrói Rodada
        // sem Game (nenhum caminho hoje) → Baralho cai no default Math.random.
        this.rng = gameSettings.rng;

        // jogadores vivos * cartas por mao + 1 pra vira. numCards e o total
        // exato consumido na rodada (darCartas + virarManilha).
        this.numCards = (this.gameOrder.length * this.round) + 1;
        // ceil: menor nº de baralhos de 40 que cobre numCards. trunc(x/40 + 1)
        // (o que estava aqui) alocava um baralho a mais quando numCards era
        // multiplo exato de 40.
        this.numBaralho = Math.ceil(this.numCards / 40);

        // Fail-fast: o monte TEM que caber a rodada exata. Se não cabe, algo
        // fora daqui está quebrado (gameOrder/round corrompido, ou a conta
        // acima) — para na construção da Rodada, antes de distribuir a
        // primeira carta, com o quadro completo.
        if (this.numBaralho * 40 < this.numCards) {
            throw new Error(
                `Rodada impossível: ${this.gameOrder.length} jogador(es) x ${this.round} ` +
                `carta(s) + vira = ${this.numCards} cartas, mas só ${this.numBaralho * 40} ` +
                `disponíveis (${this.numBaralho} baralho(s)).`
            );
        }

        this.baralho = new Baralho(this.numBaralho, this.randomShuffle, this.rng);

        this.vira = null;
        this.viraValor = -1;
        this.mesaAtiva = null;

        // Índice em gameOrder de quem inicia a vaza atual.
        // gameOrder nunca é reordenado; apenas o ponto de partida do loop muda.
        this.indiceInicial = 0;
    }

    darCartas() {
        // baralho.comprar() lança se o monte esvaziar (situação extraordinária,
        // ver Baralho.js) — não precisa checar null aqui.
        for (let j = 0 ; j < this.gameOrder.length; j++) {
            for (let i = 0 ; i < this.round; i++) {
                this.gameOrder[j].comprarCarta(this.baralho.comprar());
            }
        }
    }

    virarManilha() {
        // comprar() lança se não houver carta pra vira (ver Baralho.js) —
        // this.vira nunca é null aqui.
        this.vira = this.baralho.comprar();
        this.viraValor = (this.vira.valorInt === 9) ? 0 : this.vira.valorInt + 1;

        // A primeira vaza da rodada já pode começar com a manilha definida
        this.novaVaza();
    }

    // Cria uma mesa nova para a próxima vaza, reaproveitando o viraValor da rodada.
    // Deve ser chamado sempre que a mesa atual for concluída (todas as cartas jogadas).
    novaVaza() {
        this.mesaAtiva = new Mesa(this.viraValor);
    }

    // Retorna gameOrder reordenado a partir de quem deve iniciar a vaza atual.
    // gameOrder em si não é alterado.
    ordemDaVaza() {
        const n = this.gameOrder.length;
        const ordem = [];
        for (let i = 0; i < n; i++) {
            ordem.push(this.gameOrder[(this.indiceInicial + i) % n]);
        }
        return ordem;
    }

    registrarJogada(jogador, carta) {
        // A mesa processa a carta e já retorna o status atualizado
        return this.mesaAtiva.receberCarta(carta, jogador);
    }

    // Fecha a vaza atual: soma o steak do vencedor e define quem começa a próxima vaza.
    // Se a vaza ficou MELADO (todas as cartas se anularam), não há vencedor:
    // ninguém pontua e quem começa a próxima vaza continua o mesmo.
    finalizarVaza() {
        const vencedor = this.mesaAtiva.melhorJogada?.jogador ?? null;
        if (vencedor) {
            vencedor.steak += 1;
            this.indiceInicial = this.gameOrder.indexOf(vencedor);
        }
        return vencedor;
    }

    // Fecha a rodada: cada jogador perde hp igual à diferença absoluta entre aposta e steak
    finalizarRodada() {
        for (const jogador of this.gameOrder) {
            const diferenca = Math.abs(jogador.aposta - jogador.steak);
            jogador.hp -= diferenca;
        }
    }

    // Zera aposta e steak de todos, preparando os jogadores para a próxima rodada
    resetarApostasSteaks() {
        for (const jogador of this.gameOrder) {
            jogador.aposta = 0;
            jogador.steak = 0;
        }
    }

}

File: training/python/motor/rodada.py
# rodada.py -- porte direto de game/Rodada.js (RodadaGame): distribuir
# cartas, virar manilha, ordem da vaza, fechar vaza/rodada e a perda de hp
# (|aposta - steak|). Igual ao original, `num_baralho` cresce sozinho quando
# a rodada precisa de mais de 40 cartas (jogo grande/rodada avançada).
import math

from .baralho import Baralho
from .mesa import Mesa


class Rodada:
    def __init__(self, game_order, round_, random_shuffle, rng=None):
        self.game_order = game_order
        self.round = round_
        self.random_shuffle = random_shuffle
        # Mesmo rng do Jogo (ver jogo.py / rng.py): None => random.shuffle no
        # Baralho; um callable => Fisher-Yates deterministico.
        self.rng = rng
        # jogadores vivos * cartas por mão + 1 pra vira. num_cards é o total
        # exato consumido na rodada (dar_cartas + virar_manilha).
        num_cards = (len(game_order) * round_) + 1
        # ceil: menor nº de baralhos de 40 que cobre num_cards. (num_cards // 40) + 1
        # (o que estava aqui) alocava um baralho a mais quando num_cards era
        # múltiplo exato de 40.
        num_baralho = math.ceil(num_cards / 40)

        # Fail-fast: o monte TEM que caber a rodada exata. Se não cabe, algo
        # fora daqui está quebrado (game_order/round corrompido, ou a conta
        # acima) -- para na construção da Rodada, antes de distribuir carta.
        if num_baralho * 40 < num_cards:
            raise RuntimeError(
                f"Rodada impossível: {len(game_order)} jogador(es) x {round_} "
                f"carta(s) + vira = {num_cards} cartas, mas só {num_baralho * 40} "
                f"disponíveis ({num_baralho} baralho(s))."
            )

        self.baralho = Baralho(num_baralho, random_shuffle, self.rng)
        self.vira = None
        self.vira_valor = -1
        self.mesa_ativa = None
        self.indice_inicial = 0  # índice em game_order de quem inicia a vaza atual

    def dar_cartas(self):
        # baralho.comprar() levanta se o monte esvaziar (situação
        # extraordinária, ver baralho.py) -- não precisa checar None aqui.
        for jogador in self.game_order:
            for _ in range(self.round):
                jogador.mao.append(self.baralho.comprar())

    def virar_manilha(self):
        # comprar() levanta se não houver carta pra vira (ver baralho.py) --
        # self.vira nunca é None aqui.
        self.vira = self.baralho.comprar()
        self.vira_valor = 0 if self.vira.valor_int == 9 else self.vira.valor_int + 1
        self.nova_vaza()

    def nova_vaza(self):
        self.mesa_ativa = Mesa(self.vira_valor)

    def ordem_da_vaza(self):
        n = len(self.game_order)
        return [self.game_order[(self.indice_inicial + i) % n] for i in range(n)]

    def registrar_jogada(self, jogador, carta):
        self.mesa_ativa.receber_carta(carta, jogador)

    def finalizar_vaza(self):
        vencedor = self.mesa_ativa.melhor_jogada[1] if self.mesa_ativa.melhor_jogada else None
        if vencedor:
            vencedor.steak += 1
            self.indice_inicial = self.game_order.index(vencedor)
        return vencedor

    def finalizar_rodada(self):
        for jogador in self.game_order:
            diferenca = abs(jogador.aposta - jogador.steak)
            jogador.hp -= diferenca

    def resetar_apostas_steaks(self):
        for jogador in self.game_order:
            jogador.aposta = 0
            jogador.steak = 0
